Relative to absolute

skytreader opened the pull request:
Fixes #8.

Instead of passing around the domain, a cleaner way would be to create an object that abstracts the parsing of a document. That way, the domain can be extracted from the constructor into a field and then used by methods that need it. But I think this PR is large enough as it is for #8 so I guess I'll owe the refactor to another PR. :)

edsu commented:
I apologize for not seeing this sooner. It seems that the pull request has conflicts now. If this is something you still want to get in let me know if you can address the conflicts and I promise to apply it sooner this time!

acdha commented on microdata.py:
```diff
@@ -104,8 +113,11 @@ def json_dict(self):
-    def __init__(self, string):
-        self.string = string
+    def __init__(self, string, domain=""):
+        if string.startswith("http://") or string.startswith("https://"):
```
This feels like it's duplicating the stdlib `urljoin` semantics:

```python
>>> from urlparse import urljoin
>>> urljoin('https://example.com/foo/bar', '/baaz')
'https://example.com/baaz'
>>> urljoin('https://example.com/foo/bar', 'http://quux/baaz')
'http://quux/baaz'
```

skytreader replied:
A few questions:

1. Judging by the `.travis.yml` this project is officially on Python 3. As such I should use `urllib.parse` right?
2. Sorry I'm not sure I agree with you that this bit is duplicating `urljoin` semantics. This conditional just ensures that the string starts with either "http" or "https". `urljoin` does not do that:

```
Python 3.5.2 (default, Nov 12 2018, 13:43:14) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from urllib.parse import urljoin
>>> urljoin("github.com", "edsu")
'edsu'
```

Can you clarify if I'm missing something here?

acdha replied:
[urljoin expects its first parameter to be a URL](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin):

```
>>> from urllib.parse import urljoin
>>> urljoin("https://github.com", "edsu")
'https://github.com/edsu'
```

acdha commented on microdata.py:
```diff
@@ -115,6 +127,15 @@ def __eq__(self, other):
         return self.string
 
+    @staticmethod
+    def get_domain(url_string):
```
Perhaps this should be replaced with something using `urlparse`, which returns a handy named tuple:
```python
>>> urlparse('https://example.com/foo/bar')
ParseResult(scheme='https', netloc='example.com', path='/foo/bar', params='', query='', fragment='')
```

skytreader replied:
I used `urlparse` inside this method instead of manually splitting and joining but `urlparse` can't be a drop-in replacement for what I want to achieve here:

```
>>> urlparse("github.com/edsu/microdata/pull/32/files")
ParseResult(scheme='', netloc='', path='github.com/edsu/microdata/pull/32/files', params='', query='', fragment='')
```

acdha replied:
urlparse expects a URL:

```
>>> urlparse("https://github.com/edsu/microdata/pull/32/files")
ParseResult(scheme='https', netloc='github.com', path='/edsu/microdata/pull/32/files', params='', query='', fragment='')
```

skytreader replied:
I'm not really knowledgeable about the field here but my point is, are we sure to never encounter malformed-URLs ? I get it that urllib expects well-formed URLs but shouldn't a microdata parser be more forgiving? Hence I manually check and adapt in case the string leaves out the protocol part.

skytreader commented:
Hey there. This is not relevant to my work anymore but let me just complete the job; besides, I know others who might benefit from this. That said, I'll get back to this this weekend.

skytreader commented:
Just got back to this. Resolving conflicts and addressing comments took a while longer since I'm no longer familiar with the domain (knowledge). Anyway I see that tests are failing :(. I'll address that next but in the meantime, feel free to comment further on my clarifications.
